### DISCUSSION TITLE
Trice use on RTOS based application

### DISCUSSION THREAD

**Discussion by CMLoureiro** [rokath/trice#276] [Q&A] [answered]
Hello Thomas,

Thank you very much for making this very interesting project available to public!

I would like to use your trace project on a RTOS based application where the trace is saved into a file for later analysis.
Currently, for test purposes, I'm dumping the trace via UART.

On the test scenario I have 2 threads:
- Thread 0:  runs every 100ms to generate trace.
- Thread 1: runs every 10ms to call TriceTransfer() and  perform trace data transmission over UART ( code below)

`void triceRTOSDataTransmit(void) {
    TRICE_ENTER_CRITICAL_SECTION
    int availableTxData = TriceOutDepth();

    if (availableTxData)
    {
        HAL_UART_Transmit_DMA(&TRICE_UART,triceOutBuffer, availableTxData);
        triceOutIndex += availableTxData;
    }
    TRICE_LEAVE_CRITICAL_SECTION
}`

Questions:

On a hypothetical scenario where trace is generated faster than is consumed
- Can Trice cope with possible trace buffer overruns?
- In this situation, would it possible for the host decoder application to know that trace frames were lost?

Thank you very much for your time!
Carlos

**Comment by rokath** [comment]
2nd Question: The internal 8-bit  cycle counter is usually enabled. If trices are lost, the receiver side will detect that because the cycle counter is not as expected. There is a chance of 1/256 that the detection does not work. You can check the detection by unplugging the trice UART cable for a time.

**Reply by rokath** [reply]
Also resetting the target during transmission should display a cycle error.

**Comment by rokath** [comment, marked as answer]
1st Question: It is your responsibility to produce less data than transmittable. If this is not guarantied a data loss is not avoidable or you have to slow down the program. The double buffer as fastest solution has no overflow check. My recommendation: Make the buffer big and emit the maxDepth cyclically,  every 10 or 1000 seconds. Then you know the needed size. It is influenced by the max trice burst and the buffer switch interval.

**Reply by CMLoureiro** [reply]
Thank you very much for the clarification!
